### ISSUE TITLE
JSON Schema containing array with item reference gives Typescript error: Type '"object"' is not assignable to type '"array" | undefined'.

### ISSUE THREAD

**Issue by asnaseer-resilient** [ajv-validator/ajv#1652]
**What version of Ajv are you using? Does the issue happen if you use the latest version?**
Libraries declared in `package.json`:
```
  "dependencies": {
    "ajv": "^8.6.0"
  },
  "devDependencies": {
    "@types/node": "^15.12.2",
    "@typescript-eslint/eslint-plugin": "^4.27.0",
    "@typescript-eslint/parser": "^4.27.0",
    "eslint": "^7.28.0",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-node": "^11.1.0",
    "eslint-plugin-prettier": "^3.4.0",
    "prettier": "^2.3.1",
    "ts-node": "^10.0.0",
    "typescript": "^4.3.2"
  }
```

**Your typescript code**
Based on the **Recursion** example here: https://json-schema.org/understanding-json-schema/structuring.html
```
/* eslint-disable no-console */
// eslint-disable-next-line node/no-missing-import
import Ajv, { JSONSchemaType } from 'ajv/dist/2020';
import draft7MetaSchema from 'ajv/dist/refs/json-schema-draft-07.json';
const ajv = new Ajv();
ajv.addMetaSchema(draft7MetaSchema);

const schema: JSONSchemaType<any> = {
  $schema: 'http://json-schema.org/draft-07/schema#',

  definitions: {
    person: {
      type: 'object',
      properties: {
        name: { type: 'string', nullable: false },
        children: {
          type: 'array',
          minItems: 0,
          items: { $ref: '#/definitions/person' },
          nullable: false
        }
      },
      required: ['name']
    }
  },

  type: 'object',

  properties: {
    person: { $ref: '#/definitions/person' }
  }
};

const validator = ajv.compile(schema);

const validData = {
  person: {
    name: 'Elizabeth',
    children: [
      {
        name: 'Charles',
        children: [
          {
            name: 'William',
            children: [{ name: 'George' }, { name: 'Charlotte' }]
          },
          {
            name: 'Harry'
          }
        ]
      }
    ]
  }
};

if (validator(validData)) {
  console.log(validData);
} else {
  console.log(validator.errors);
}
```

**Typescript compiler error messages**
```
    return new TSError(diagnosticText, diagnosticCodes);
           ^
TSError: ⨯ Unable to compile TypeScript:
src/issue.ts:13:7 - error TS2322: Type '"object"' is not assignable to type '"array" | undefined'.

13       type: 'object',
         ~~~~

    at createTSError (/Users/anjum/Documents/workspace/resilient/playground/json-schema-validator/node_modules/ts-node/src/index.ts:587:12)
    at reportTSError (/Users/anjum/Documents/workspace/resilient/playground/json-schema-validator/node_modules/ts-node/src/index.ts:591:19)
    at getOutput (/Users/anjum/Documents/workspace/resilient/playground/json-schema-validator/node_modules/ts-node/src/index.ts:921:36)
    at Object.compile (/Users/anjum/Documents/workspace/resilient/playground/json-schema-validator/node_modules/ts-node/src/index.ts:1189:32)
    at Module.m._compile (/Users/anjum/Documents/workspace/resilient/playground/json-schema-validator/node_modules/ts-node/src/index.ts:1295:42)
    at Module._extensions..js (internal/modules/cjs/loader.js:785:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/anjum/Documents/workspace/resilient/playground/json-schema-validator/node_modules/ts-node/src/index.ts:1298:12)
    at Module.load (internal/modules/cjs/loader.js:641:32)
    at Function.Module._load (internal/modules/cjs/loader.js:556:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:837:10)
```

**Describe the change that should be made to address the issue?**
I do not know. However, I spotted that if I change **#/definitions/person/properties/children/nullable** to **true** then the error disappears in VSCode but the Typescript compiler still gives the same error.

**Are you going to resolve the issue?**
No because I do not know why this is happening.

**Comment by epoberezkin** [comment]
I think nullable: false is considered invalid to be used in the schema with JSONSchemaType, it would be no-op. You can just remove it.

**Comment by asnaseer-resilient** [comment]
@epoberezkin If I remove all the **nullable** fields it still fails with the same error.

**Comment by epoberezkin** [comment]
I actually don't think `JSONSchemaType<any>` is supposed to work, it expects a particular data type/interface. If you need the type for any JSON Schema, there is SomeJSONSchema (I think)

**Comment by asnaseer-resilient** [comment]
I created **Person** and **PersonHolder** interfaces for schema like this:
```
...
interface Person {
  name: string;
  children?: Person[];
}

interface PersonHolder {
  person: Person;
}

const schema: JSONSchemaType<PersonHolder> = {
  $schema: 'http://json-schema.org/draft-07/schema#',

  definitions: {
    person: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        children: {
          type: 'array',
          minItems: 0,
          items: { $ref: '#/definitions/person' }
        }
      },
      required: ['name']
    }
  },

  type: 'object',

  properties: {
    person: { $ref: '#/definitions/person' }
  }
};
...
```
Even with this I get the same error.

**Comment by asnaseer-resilient** [comment]
I tried to use **AnySchemaObject** and the code compiles and runs correctly with this.

So there must be some Typescript errors somewhere in **JSONSchemaType** or I have made some other type of basic mistake in my usage of this library.

**BTW**: I am a complete new comer to this library.

**Comment by epoberezkin** [comment]
I can reproduce.

The schema type does not "see through" references, so it should allow any schema in definitions. For some reason though it seems to match one of the branches that expects the array type.

This schema fails on type: "object":

```javascript
const schema: JSONSchemaType<number> = { // can be any type
  type: "number",
  definitions: {
    person: {
      type: "object", // fails here, requires type to be "array" or undefined
      properties: {
        name: {type: "string"},
        children: { // removing "children" property (or making it non-array type) makes it pass
          type: "array"
        }
      },
      required: ["name"]
    }
  },
}
```

but this schema is ok:

```javascript
const schema: SomeJSONSchema = {
  type: 'object',
  properties: {
    name: { type: 'string' },
    children: {
      type: 'array',
    }
  },
  required: ['name']
}
```

even though "definition" is defined as `{[Key in string]?: UncheckedJSONSchemaType<Known, true>}` and "SomeJSONSchema" is `UncheckedJSONSchemaType<Known, true>` 

Something weird is going on here...

cc @erikbrinkman

**Comment by epoberezkin** [comment]
I think I understand the problem - JSONSchemaType seems to require that optional properties have an explicit nullable true in them, although the way this problem comes out is very confusing... So this compiles:

```
const schema: JSONSchemaType<number> = {
  type: "number",
  definitions: {
    person: {
      type: 'object',
      properties: {
        name: { type: 'string' },
        children: {
          type: 'array',
          nullable: true
        }
      },
      required: ['name']
    }
  },
};
```

**Comment by asnaseer-resilient** [comment]
@epoberezkin You said that the generic parameter type of `JSONSchemaType` "can be any type" in here: `const schema: JSONSchemaType<number> = { // can be any type`

If thats the case, then what is the purpose of that generic type parameter?

**Comment by epoberezkin** [comment]
I mean that the error happens irrespective of which type you use there, it’s specifically related to definitions. This type parameter should be your dat type.

**Comment by erikbrinkman** [comment]
The problem actually seems ultimately to have more to do with the `$ref:`. `JSONSchemaType` doesn't recognize a standalone ref as a valid type, e.g. `const x: SomeJSONSchema = { $ref: "#" }` fails. In your example, if you replace it with a simple schema, e.g. `{ type: "number" }` you get a new error, which is valid, you still need to add `required: ["person"]` to the top level schema, and then everything works.

@asnaseer-resilient,  since all this does is help with type checking, and definitions aren't type checked in `JSONSchemaType` (they're only checked if they structurally match any schema definitions), I recommend if you have a similarly definition based type, for now you just ignore it with `AnySchemaObject` and move on. (Note, if your schema is simple / fits within the JTD spec, the JTD types do type definitions and refs).

@epoberezkin A possible fix for this to also allow a catch-all ref check similar to the union types , e.g. right after:
https://github.com/ajv-validator/ajv/blob/f8708387a5be62fbd0becc0cedfc580021a5c5da/lib/types/json-schema.ts#L32-L39
just add
```
    | { $ref: string }
```
It's not super elegant, but it's the only way I can see to treat $refs appropriately.

My main question is how this plays with `JSONSchemaType` overall. Does it make sense (according to spec) to allow refs tagged onto other things e.g. 
```
{
    $ref: "#/array",
    type: "array",
    minItems: 3 // i.e. overriding something from the ref
}
```
or should they always be standalone. If the latter, then this:
https://github.com/ajv-validator/ajv/blob/f8708387a5be62fbd0becc0cedfc580021a5c5da/lib/types/json-schema.ts#L133
should probably also be removed (although the `[keyword: string]: any` also makes the removal somewhat meaningless. Actually, on a side note, I'm not sure why that's there)

**Comment by asnaseer-resilient** [comment]
Thanks all for your responses to this. FYI, this is the actual schema I defined in the end for my application using `JSONSchemaType` (I am not sure if JTD can be used as I am unclear on its support for unions).
```
import { AnySchemaObject } from 'ajv/dist/2020';

export const UI_PERMISSIONS_META_SCHEMA: AnySchemaObject = {
  $schema: 'http://json-schema.org/draft-07/schema#',

  definitions: {
    uuid: { type: 'string', minLength: 36, maxLength: 36 },
    groupName: { type: 'string', minLength: 1, maxLength: 32 },
    description: { type: 'string', maxLength: 100 },
    label: { type: 'string', minLength: 1, maxLength: 32 },
    effect: { type: 'string', enum: ['ALLOW', 'DENY'] },
    action: { type: 'string', minLength: 1 },
    resources: { type: 'array', minItems: 0, items: { type: 'string' } },
    PermissionStatement: {
      type: 'object',
      properties: {
        effect: { $ref: '#/definitions/effect' },
        action: { $ref: '#/definitions/action' },
        resources: { $ref: '#/definitions/resources' }
      },
      required: ['effect', 'action', 'resources']
    },
    UiPermissionStatements: {
      type: 'object',
      properties: {
        label: { $ref: '#/definitions/label' },
        statements: {
          type: 'array',
          minItems: 1,
          items: {
            $ref: '#/definitions/PermissionStatement'
          }
        }
      },
      required: ['label', 'statements']
    },
    UiPermissionSpecificStatements: {
      type: 'object',
      properties: {
        uuid: { $ref: '#/definitions/uuid' },
        label: { $ref: '#/definitions/label' },
        statements: {
          type: 'array',
          minItems: 1,
          items: {
            $ref: '#/definitions/PermissionStatement'
          }
        }
      },
      required: ['uuid', 'label', 'statements']
    },
    UiPermissionPolicyOption: {
      type: 'object',
      properties: {
        label: { $ref: '#/definitions/label' },
        description: { $ref: '#/definitions/description' },
        permissionOptionType: { enum: ['Policy'], type: 'string' },
        policyId: { type: 'string', pattern: '^V#V[0-9]{4}_[0-9]{2}_[0-9]{1}#P#[a-zA-Z][a-zA-Z_]*$' }
      },
      required: ['label', 'description', 'permissionOptionType', 'policyId']
    },
    UiSpecificPermissions: {
      type: 'object',
      properties: {
        label: { $ref: '#/definitions/label' },
        items: {
          type: 'array',
          minItems: 1,
          items: { $ref: '#/definitions/UiPermissionSpecificStatements' }
        }
      },
      required: ['label', 'items']
    },
    UiPermissionSetOption: {
      type: 'object',
      properties: {
        label: { $ref: '#/definitions/label' },
        description: { $ref: '#/definitions/description' },
        permissionOptionType: { enum: ['PermissionSet'], type: 'string' },
        permissionSet: {
          type: 'object',
          properties: {
            uuid: { $ref: '#/definitions/uuid' },
            any: { $ref: '#/definitions/UiPermissionStatements' },
            specific: { $ref: '#/definitions/UiSpecificPermissions' }
          },
          required: ['uuid', 'any', 'specific']
        }
      },
      required: ['label', 'description', 'permissionOptionType', 'permissionSet']
    }
  },

  type: 'object',
  properties: {
    groupName: { $ref: '#/definitions/groupName' },
    label: { $ref: '#/definitions/label' },
    description: { $ref: '#/definitions/description' },
    options: {
      type: 'array',
      minItems: 1,
      items: {
        anyOf: [{ $ref: '#/definitions/UiPermissionPolicyOption' }, { $ref: '#/definitions/UiPermissionSetOption' }]
      }
    }
  },
  required: ['groupName', 'label', 'description', 'options']
};
```

**Comment by erikbrinkman** [comment]
@asnaseer-resilient [JTD](https://ajv.js.org/json-type-definition.html) only supports tagged unions, but it looks like your schema does have a tagged union, so that should work and provide better type guarantees than `JSONSchemaType`, although I don't want to go into the specifics here. However, JTD is only a structural definition, so it doesn't allow for the other validation that you do here like regex or min / max length for lists or strings. What you end up choosing is up to you, but the rough trade off is that JTDs schema allows easier / better typescript static verification that the schema matches the type returned, but JSONSchema is much more expressive, so if you trust what you wrote, it support more complex verification.

**Comment by asnaseer-resilient** [comment]
@erikbrinkman Thank you very much for your insights - much appreciated. For now I think I will stick with `JSONSchemType` as it provides more expressive validations.

**Comment by erikbrinkman** [comment]
@asnaseer-resilient One more little comment because I stumbled onto when looking over the tests. They did the smart thing of using:
```
const schema: JSONSchemaType<MySchema> & {
  definitions: {
    a: JSONSchemaType<...>,
    ...
  }
} = { ... }
```
to also check the definitions. Note, it won't make sure that the references to those definitions match the type, just that the individual definitions also type check correctly. Still it will provide a little more guarantees in certain circumstances.

**Comment by epoberezkin** [comment]
while it's a limitation that definitions / refs are not typed, the issue still remains as per https://github.com/ajv-validator/ajv/issues/1652#issuecomment-868973633

This schema should be valid against this time, I still don't follow why it is not...

@erikbrinkman

**Comment by erikbrinkman** [comment]
Looked over this again, and there are number of things going on still, depending on which MWE you're dealing with.

The fix for #1664 should deal with needed nullable for an optional type. The other problems are
1. is `items` necessary for array type? then it won't allow you to just declare `{ type: "array" }`. I think items should be required, but it does prevent some of the schemas above.
2. is `{ $ref: "..." }` a valid type. Currently it's hacked into `properties` to be allowed
   https://github.com/ajv-validator/ajv/blob/a04657014022c9ecb9ae0738b2ac07cb0d76cf3f/lib/types/json-schema.ts#L151-L153
   but not for items, or any other top-level type, e.g. `{ definitions: { a: { ... } }, $ref: "#/definitions/a" }` won't work. The "best" resolution for all of this issues is to probably remove `{ $ref: string }` from above, and add it here
   https://github.com/ajv-validator/ajv/blob/a04657014022c9ecb9ae0738b2ac07cb0d76cf3f/lib/types/json-schema.ts#L32-L40
   loosely allowing any type to be specified as a `$ref`.

**Comment by epoberezkin** [comment]
> is items necessary for array type?

good question. Arguably it’s very rare that you need [any] type in the schema and if you do you can still have items: {}.

I’d say if it’s easy to make items: {} optional then it’s good to do it, but if it’s lots of type-foo needed for it we can just add the line in the docs…

> is { $ref: "..." } a valid type

yes. It’s not too uncommon to have items defined as $ref

**Comment by epoberezkin** [comment]
fixed in v9

**Comment by StefanR23** [comment]
2024 and still no recursive patterns? see also #1728. Or is this somehow doable in another way?

**Comment by erikbrinkman** [comment]
@StefanR23 recursive types are still not supported for JSONSchemaType. In principle it may doable, but there are a lot of edge cases with JSONSchema that make it hard to test in typescript. A key one is static url parsing, and then resolving if it's local or remote. More critically, the issue of determining that recursion is necessary and how it might apply. In all likelihood it would need to have the same structure as JTDSchemaType which requires to manually specify the type of all the $defs too.

More to the point, the issue you reference seems to actually be due to three issues:
1. missing `required`, unfortunately due to the way typescript type checks, it's hard to make sure this appears obviously
2. mistyped $defs, which should `SomeJSONSchemaType` in it's true form
3. JSONSchemType doesn't want to have a single $ref at the root level

If you [fix all of these](https://www.typescriptlang.org/play/?#code/PQKgBApgzgNglgOwC4FoAmcoEMBGMJgACSAngA7QDGATnGatPMsAgPYoQC29JKiSEagDMslAiGAAoUhTABlJLUpIAcgFcYMAMIALCJQDWUAOrUsZCtQA8KrJwIQAHgIRooYKIsQBzADRgAFXIIAD4wAF4wNVcIIUQINEhnCFd3BA0YSTBssAB+MAADTyVVDN19IzBONU8wHAJFNQJEMCQoSlYEOO9W1iioAgASAG9bewBfAqycgC5A4MlpYLAAVQQ4ToDWAElkQQHlDYQrFbDIgAoVpJc3MCwEEjywc4B9OZWASgiwgDdWOEScwQEB+gi+Thu7lec0QQkEYG2X3Cv3+aGm2Xy23RYCBIMEiycZFY1CQrWWclY9gAUnIAPIqOSUPScLARVYIJkVBI0+mM5lYIIUKwAaTYAHcEP5GqFFjICGtOYYEgAFLAkuBYGB8rhYKwBM5gVXqzUnDl6JVoHkMzkswUQPVS6hNEIhAmOIkksmyI1IDVam26-VshQldSacqGExmCyCKwAIh9fu1LLj-gV5oMKrVvs1ycDLtlyytdr11xSt2KPn82ygic1ZdSdVYrHw9wNNbrMAbt2l2PyATAAB8ojE4sC0TkcYFFvxBCIxGB1Jx6tRhRASGLibdhtjOIg4NVOLkgWpl-jJyzHAfT8ewOkz9RsU5KDAanBQQBZfeH2-3ldPxwXzfT8sCvH8TwfXcNF9Mh8FpIRf1Pf9JyEYkWSQW9KwQbxJHGGc9mEUQCBDHw1w3Ld3B3C9EAAGRSbwkB0RDIIvUC6OwxjmOQnIyCwJABGoBBMK8bDsVQ6h0OE2hRLwpZZHTLlLTpa1+RLAJq1rbM-W7dwcGbVsEANc5sWHYBgFaPQBlaTcRyOdxNRgVgxTuagcDgRQ1UeaI7LAVghC9aBsSoyccnuEh4LmagICwNBOhgR4FItK08zUjTOxCABtABdbE8MnYdgpC3zgQisAopiuKErNRTkoDVKEU040YEynLJzynIzIszBbM6O5NCc9xxLAYYArmc4yFoPdfVBD5srAcYPFIfAPADKATOeQqQrlSLotihB4ueAcIXLNIkPPIqQvyYtgnjP9BDjIcwDjWdvHutKtM1V0LpCuYjuSRssJw77Jyu5SSzjQHUwa9LsW+36dKbFtooQWGLtB+lwb0pH7ihjsPua1GitxUFqFm1qcgWgAydkji2XYBIOX1OisQnAgRu7H2+-IlxXMjN2oNxWfh46AZEoG0fkMW+YooW2ZF24sYM1n8k6xh+HQTBcHwFBgWcFAmAIYhgnaWh6A4WB+GAHB7hQOU1u+4Z2qJu88U5nIQg+bFOsY7qOhJ9w4zYCTNQeu2AG4wHcgBydwYCwWooD6O2aDoUlHNYSokD6dywDiahaiGwRqGJe2cmHc5zj+yE7zOt2ivyTbvu2sBrqFQPa4e4dnr2V7qFxxqcwJ76qcXWvpYF0vnark7Fuk8X6+G1nJ2b1v7QhsX+5h4ewGpkjsPHwW4bl-6Ff05HlcX4GchXsGbrjRXkc3-Gvoup2fuP6vyr2g6MrC-wADpAFhWyuTBejcLqdStGAAAgg8PMudiStDULBCAS9r7BDmKveMaozAkCfk1F+V93JcCgHMcBV8v6VTABlYUkcEBgEzCQPygQsooFvIlTMSleR1RugEGhWV-AiBgAMMI1MwxxzwPaPhwosqEOBiPEa+AOI6F+hlOMSiGI6DjFleaaDsh7gQNsAQnBSGBDURoxi2jWYj3OCNS8RiSGqPUfRSxOiFoFTuGgDATMECagcSYuYQirLjE9kfaejZKH7UeMA0Bl1L5X1Gi3W+bccFYDwe9AhejI7GNMRw7kykUq8IygABgEbnTUIiskdGQFgRAUB2HVQtJ2PMeoSmyKyQY-x9SILcSvvYnJXFzp9MQFoToSBakIG6TXFiCTLyjJqXUwZddgbeQAI5NC6beXsCTPHeKOH4gZxMhkhTfpOYWJ93AACV9BbisIDfwYU5Eg3iVfSBykYFwIDAg6gZUbkTzuK4MAGBDidDVHAQKOzOpxiiusuAUU0APW6mwUkrB6D7K7PUSgWAajNFJN1PyLhvneFYPxFIWSoUTVRYIX00AHpqgIKinx9YhqcGJAQaplBMAEGBUy6gjx2j8knsDClfEBIIGVMXSwNKoB0sBVi+h9RioHRxYkMU7kdB0IEvOaAFk+JgB0FgUEs8fB0LQE4LJN8MZ31YDgAAVvoJA+DB5PIujFPZoKYASqpeqaAt4H73EenkrhKkdRqQyoDMpQSZQ7OiCCTUag+IqkldS8FUz-X0OHEG2qqkikRsERU6NCTKVStTbePGTUEbbISfkTsppFScK9SW6ALT9QuudkGxtKbm0Bj1G2kKvFSWCU7T6qZ1yOgCzuWLNMjTOHZtDbmsWkaC0Fh2cW6lJAxi+rmLSKadaMz5O4TmoU9zynCNCP4OMco4yiOGgFW86854PVORdM1FAYgclLWQ6htCWiMOYQELKt5IkHXOH+-yAQyaBpnVmJqLawjPqKq+8sKQkDXNhfC28I0aF0IYeuf9gGdoVSic8MDgRIMIZCkh99SA8xTKwz++hpGAMNPrTBwecHdE7IMcO6VSyOmgR45+6ZvSTk72eOWwelanQEChTCtQcKEiIrSCSl2YgoDYD5ZHfy0d7J3k6CgXiFa5RgDVYxPTvz0MJFwyQdwnV6VgHgJ4KzOcqHxwGCSdwjE9WMvRWAeVyK6gEEcwINA-8wCyYgJZhFAUgWsB1QF7wCazB7AaAa0kDkLPyfhWANdI6skN0ywptAhHv6PFA3h8D5GslzAykGtDWWEgfi4CuKAvadHy3cBlYEJNYlc1vXJorQHdpULq5FhraAmsPla-qOaFGzn9bG0VkrI3oNoHq0VybLXe2zdCRdc51d0iaAvuQraGCklWrbodmATq-R9snFdrWEA5hVqKhRo5pMxMncnA5eCQ2iMHQ7fjOD2UslhV+8t4jgPYM9pm71i6nQIDg9+f9qqrG0DNJhy1LJcAELvFWxj-kvaLV6CEnjtHBOdRE52RAM9LH93o6B5jrJyK6eKQpyyKnr9PZfGpptDKjD+bFeNdhLKcwwrYkGACKSPgJdRVx8L+eYBBhmqEKOv5aBJ1z38NEAw4pDLYhV-uHxavx0a5PTrvXrpZKEmJKSYzc7bS8INCRZQ4iIxGFMOYSwLNJxxgdwKYIqZsRZoKTwoU6lT2VNdHJAgoonL0PCOtLDjC5gRrmHHiUnGy7fr1wAwBGeEAgPWgXkH+VhPHOHIDda6b1pXcLPJfHyaR0cciHzhj1n8NsLGiHw987w-8PzWem94jHutJkR7R6I1Bhy9T2LXR1v3S25i4J7thOgyRBd6UcMGYoxe9jNiBMTfpV5iD5ODtR-U1wckNH4zo2oubf2L2tkbecNMdYbeaIhvxwI2kTo-I3X4R09cJ+cKsWE3QPQ7dlh1t4UH984n8N8vBXcygd9PcYxrAD9oDGtmt9hT8cg79xtYDptXQb9lgR9JF4CRwv8rMOtpxnkTsHtJFntpNWZqlPBEJNBwtzIwo9MEAdYMgwAfh40CBNwNBEgRA4AuwChWCkATxNACh-ACgrsCgTNWBRDykJCAVHgSU9AflBDXxmh6FpDWYUgbwIcQMBxhwrtINOpFCMhlDqhahFVCR4BOUkADpEAgIzVEgWg4wTDOAHohpbC5Degcs3NWYVdsUYAMJ4ZLCMhcpsQyFWYGD8Bbwo0WDOg2DfpjD7w-tSszE4dsgIjoJbwAh4jFhwCl9ZxCIFxtgaJMBSRNomQJC0AopScEQ6jPBsow5FhZJpCVp+RMFQ8j17Raj6iDRNpm44wbV7VlAoZOCup3AoAdBVCYA0AEAo5SRFUdDUFJxp9Yg5g4wABiYAZXWIKAYAYLXA7IU41XRI76YLO4hJSY6Yh1K44GXLaVR4nZbIJo1Y1or4745eM7OMVJdJLJb6YhAJW9PYoQA444m484y4+aXwcEi6ZIp7VoZgwE7IObCmFEnZAbeFGrARUHLxdydFFfUxNI4eeI8YMOIAA) you no longer get compile errors.

Note, I didn't actually fix 1, I just made a bad schema to handle it, or 2, because the current implementation actually defines SomeJSONSchema wrong, as noted in another Issue. The hack of passing in the Known union doesn't actually expand correctly to any possible JSONSchemaType, and as a result, you get a lot of weird bugs.

Also note that even if you did do all of these, the schema isn't actually checked, it's just assumed to be correct.

It's probably also worth pointing out that there's a 9.0 draft that fixes some other issues, notably, this field shouldn't be nullable, as that would imply the value can also be null, not just missing.

### What can you do?

This was a lot of saying why it doesn't work, but not much of what can you do in the meantime. Unfortunately, I don't have a great solution. JSONSchema is really complicated, and this code has some tech debt in it that is hard to clean up for compatibility reasons.

For simple types just use jtd. it's a much friendly schema for just type checking. The referenced issue can be rendered as:
```
const schema: JTDSchemaType<IList, {"list": IList}> = {
    ref: "list",
    definitions: {
        list: {
            optionalProperties: {
                children: {
                    elements: { ref: "list" },
                }
            },
        }
    }
};
```
which type checks the whole thing accurately. You can also use JTDDataType, but that doesn't work well here since it can be a little finicky with objects that only have optional fields, partially due to the way that typescript works, although this might be a bug.

If you really need JSONSchema, you can always try rolling your own, after all, AJV doesn't actually use the type, that's just for developer help. In the example above you can make a similar type with something like:
```
type InnerType<T, D extends Record<string, unknown>> = (
    T extends readonly (infer I)[] ? {
        type: "array"
        items: InnerType<I, D>
    } : {
        type: "object"
        properties: {
            readonly [K in keyof T]-?: InnerType<T[K], D>
        },
        required: readonly (keyof T)[],
        additionalProperties?: boolean,
    }
) | {
    [K in keyof D]: T extends D[K] ? K extends string ? { $ref: `#/\$defs/${K}` } : never : never
}[keyof D]

type MyJSONSchemaType<T, D extends Record<string, unknown>> = InnerType<T, D> & {
    $defs: {[K in keyof D]: InnerType<D[K], D>}
}

export interface IList {
    children?: IList[];
}

const schema: MyJSONSchemaType<IList, {"list": IList}> = {
    $ref: "#/$defs/list",
    $defs: {
        list: {
            type: "object",
            properties: {
                children: {
                    type: "array",
                    items: { $ref: "#/$defs/list" },
                }
            },
            required: [],
            additionalProperties: false
        }
    }
};
```

This actually type checks the recursive portion, but will obviously get more complicated as you use more features from JSONSchema.

**Comment by StefanR23** [comment]
@erikbrinkman  Tanks for the snippets.
But I'm getting lost in my more complex project: e.g.  items: [oneOf: [...] ] etc. I need more time to get into infer the types.

Meanwhile in parallel I found a workaround. Not the best, but fine for the moment to get code completion to a certain level etc.

```
import AJV, { JSONSchemaType } from "ajv/dist/2020";

const ajv = new AJV({ allErrors: true });

export interface Tree {

    data: string
    children?: Tree[];
}

// works only as embedded and not as standalone schema
const treeSchema : JSONSchemaType<Tree> = {
    type: "object",
    properties: {
        data: {type: "string"},
        children: {
            type: "array",
            items: { $ref: "#/$defs/tree" } as unknown as JSONSchemaType<Tree>,
            nullable: true
        }
    },
    required: ["data"],
    additionalProperties: false
};

interface Document {
    title:string;
    tree: Tree
}

const documentSchema : JSONSchemaType<Document> = {
    type: "object",
    properties: {
        title: { type: "string"},
        tree:  {  $ref: "#/$defs/tree" }
    },
    additionalProperties: false,
    required: ["title", "tree"],
    $defs: {
        tree: treeSchema  as unknown as JSONSchemaType<Tree>
    }
};


const validator = ajv.compile<Document>(documentSchema);

const subject : Document = {
    title: "title",
    tree: {
        data: "root",
        children: [
            {
                data: "child",
                children: [
                    {
                        data: "grandchild",
                    }
                ]
            }
        ]
    }
}

if(validator(subject)){
    console.log("fine")
} else{
    console.log(JSON.stringify(validator.errors))
}

```
